keadm/cmd/keadm/app/cmd/edge/batchprocess.go

```diff
@@ -0,0 +1,475 @@
+		if err := os.MkdirAll(binOutputDir, os.ModePerm); err != nil {
+			return errors.Errorf("failed to create directory %s: %v", binOutputDir, err)
+		}
+
+		if err := extractTarGz(packagePath, binOutputDir); err != nil {
+			return err
+		}
+		klog.Infof("Extracted keadm package for keadm-%s-linux-%s.tar.gz to %s", cfg.Keadm.KeadmVersion, arch, binOutputDir)
+	}
+	return nil
+}
+
+// download keadm package
+func downloadKeadmPackages(cfg *common.Config) error {
+	for _, arch := range cfg.Keadm.ArchGroup {
+		var url string
+		if cfg.Keadm.Download.URL == nil {
+			url = fmt.Sprintf("%s/%s/keadm-%s-linux-%s.tar.gz", defautKeadmDownloadURL, cfg.Keadm.KeadmVersion, cfg.Keadm.KeadmVersion, arch)
+		} else {
+			url = fmt.Sprintf("%s/keadm-%s-linux-%s.tar.gz", *cfg.Keadm.Download.URL, cfg.Keadm.KeadmVersion, arch)
+		}
+		outputDir := filepath.Join(packageDir, arch)
+		if err := os.MkdirAll(outputDir, os.ModePerm); err != nil {
+			return errors.Errorf("failed to create directory %s: %v", outputDir, err)
+		}
+		outputPath := filepath.Join(outputDir, fmt.Sprintf("keadm-%s-linux-%s.tar.gz", cfg.Keadm.KeadmVersion, arch))
+
+		// decompressed into target directory
+		binOutputDir := filepath.Join(binDir, arch)
+		if err := os.MkdirAll(binOutputDir, os.ModePerm); err != nil {
+			return errors.Errorf("failed to create directory %s: %v", binOutputDir, err)
+		}
+
+		// attempt extract
+		klog.Infof("Attempting to extract keadm for %s to %s", arch, binOutputDir)
+		if err := extractTarGz(outputPath, binOutputDir); err != nil {
+			klog.Warningf("Failed to extract file %s, will attempt to download: %v", outputPath, err)
+
+			// download keadm
+			klog.Infof("Downloading keadm for %s from %s to %s", arch, url, outputPath)
+			if err = downloadFile(url, outputPath); err != nil {
+				return err
+			}
+
+			// attempt extract again
+			klog.Infof("Re-attempting to extract keadm for %s to %s", arch, binOutputDir)
+			if err = extractTarGz(outputPath, binOutputDir); err != nil {
+				return errors.Errorf("failed to extract file after download: %v", err)
+			}
+		}
+
+		klog.Infof("Downloaded and extracted keadm for %s to %s", arch, binOutputDir)
+	}
+	return nil
+}
+
+func downloadFile(url, outputPath string) error {
+	cmd := exec.Command("curl", "-L", "-o", outputPath, url)
+	if err := cmd.Run(); err != nil {
+		return errors.Errorf("failed to download file from %s: %v", url, err)
+	}
+	return nil
+}
+
+func extractTarGz(tarFile, destDir string) error {
+	cmd := exec.Command("tar", "-xzvf", tarFile, "-C", destDir)
+	if err := cmd.Run(); err != nil {
+		return errors.Errorf("failed to extract tar.gz file: %v", err)
+	}
+	return nil
+}
+
+// batch process nodes
+func batchProcessNodes(cfg *common.Config, logWriter *bufio.Writer) error {
+	var wg sync.WaitGroup
+	sem := make(chan struct{}, cfg.MaxRunNum)
+	for _, node := range cfg.Nodes {
+		wg.Add(1)
+		go func(node common.Node) {
+			defer wg.Done()
+			sem <- struct{}{}
+			defer func() { <-sem }()
```

Comment:
What is the use of channel here?

Reply:
The channel of sem is to limit the number of concurrent processing nodes, which is controlled by MaxRunNum